apps/sms/views/shared/js/system_message_handler.js

```diff
@@ -0,0 +1,341 @@
+      ]).then(([titleL10n, bodyL10n, icon]) => {
+        // Force the helper to leave close behavior as it is because it should
+        // be handled in Utils.closeNotificationsForThread later.
+        return NotificationHelper.send(titleL10n, {
+          icon: icon,
+          bodyL10n: bodyL10n,
+          tag: 'threadId:' + message.threadId,
+          data: { id: message.id, threadId: message.threadId },
+          closeOnClick: false
+        });
+      }).then((notification) => {
+        notification.addEventListener(
+          'click', () => this.onNotificationClicked(message)
+        );
+
+        // Close notification if we are already in conversation view and app
+        // becomes visible.
+        if (document.hidden && isInConversation) {
+          Utils.onceDocumentIsVisible().then(() => notification.close());
+        }
+      });
+    },
+
+    /**
+     * Fires once notification is clicked.
+     * @param {{ id: string, threadId: string }} notification Notification
+     * arbitrary data.
+     * @returns {Promise.<void>} Promise that is resolved once notification
+     * click is handled.
+     * @private
+     */
+    onNotificationClicked({ id, threadId }) {
+      return navigator.mozApps.getSelf().then((app) => {
+        app.launch();
+
+        // If we're currently in the target conversation, just do nothing.
+        if (Navigation.isCurrentPanel('thread', { id: threadId })) {
+          return;
+        }
+
+        //Validate if message still exists before opening message thread
+        //See issue https://bugzilla.mozilla.org/show_bug.cgi?id=837029
+        return Utils.onceDocumentIsVisible().then(
+          () => MessageManager.getMessage(id)
+        ).then(
+          (message) => Navigation.toPanel('thread', { id: message.threadId }),
+          () => Utils.alert('deleted-sms')
+        );
+      });
+    },
+
+    /**
+     * Handles "sms-received" system message.
+     * @param {MozMmsMessage|MozSmsMessage} message Received message.
+     * @returns {Promise.<void>} Promise that is resolved once message is fully
+     * handled.
+     * @private
+     */
+    onSmsReceivedSystemMessage(message) {
+      // Acquire the cpu wake lock when we receive an SMS.  This raises the
+      // priority of this process above vanilla background apps, making it less
+      // likely to be killed on OOM.  It also prevents the device from going to
+      // sleep before the user is notified of the new message.
+      //
+      // We'll release it once we display a notification to the user.  We also
+      // release the lock after 30s, in case we never run the notification code
+      // for some reason.
+      this.requestWakeLock();
+
+      var handleReceivedMessagePromise;
+      if (message.messageClass === 'class-0') {
+        handleReceivedMessagePromise = this.onMessageZeroClassReceived(message);
+      } else {
+        handleReceivedMessagePromise = SilentSms.checkSilentModeFor(
```

Comment:
I keep ```SilentSms.checkSilentModeFor``` in this method instead of ```onMessageNonZeroClassReceived``` because I feel it's a separate thing, but tell me what you think!